### ISSUE TITLE
Time::Piece reports timezone incorrectly

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9684]
Migrated from [rt.perl.org#63914](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=63914) (status was 'resolved')

Searchable as RT63914$

**Comment by p5pRT** [comment]
### From @ig3

#### Created by ian@alula.local



$ ~/localperl/bin/perl \-MTime&#8203;::Piece \-e '$t = Time&#8203;::Piece\->localtime; 
print $t\->strftime\("%z"\) \. "\\n"'
\+0000

but

$ ~/localperl/bin/perl \-MTime&#8203;::Piece \-e '$t = Time&#8203;::Piece\->localtime; 
print $t\->tzoffset \. "\\n"'
46800

and

$ date
Mon Mar 16 16&#8203;:09&#8203;:57 NZDT 2009

The attached patch seems to solve the problem on my system and all tests 
are successful with otherwise default configuration on CentOS 5\.2

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low
    module=Time::Piece

Site configuration information for perl 5.11.0:

Configured by ian at Sat Mar 14 11:06:39 NZDT 2009.

Summary of my perl5 (revision 5 version 11 subversion 0) configuration:
  Commit id: df3467db16fc4437950db94f0d3ad7ff8b2c9e46
  Platform:
    osname=linux, osvers=2.6.18-92.el5, archname=i686-linux
    uname='linux alula.local 2.6.18-92.el5 #1 smp tue jun 10 18:49:47 
edt 2008 i686 i686 i386 gnulinux '
    config_args='-Dprefix=/home/ian/localperl'
    hint=previous, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-DDEBUGGING -fno-strict-aliasing -pipe 
-fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE 
-D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector 
-I/usr/local/include -DDEBUGGING -fno-strict-aliasing -pipe 
-fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE 
-D_FILE_OFFSET_BITS=64'
    ccversion='', gccversion='4.1.2 20071124 (Red Hat 4.1.2-42)', 
gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.5.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -L/usr/local/lib 
-fstack-protector'

Locally applied patches:
    PERL_GIT_UNPUSHED_COMMITS           /* do not remove this line */
    PERL_GIT_UNCOMMITTED_CHANGES        /* do not remove this line */


@INC for perl 5.11.0:
    /home/ian/localperl/lib/site_perl/5.11.0/i686-linux
    /home/ian/localperl/lib/site_perl/5.11.0
    /home/ian/localperl/lib/5.11.0/i686-linux
    /home/ian/localperl/lib/5.11.0
    .


Environment for perl 5.11.0:
    HOME=/home/ian
    LANG=en_NZ.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    
PATH=/usr/local/java/jdk/bin:/usr/local/java/ant/bin:/usr/kerberos/bin:/usr/local/bin:/usr/bin:/bin:/usr/X11R6/bin:/usr/local/Trolltech/Qt-4.4.3/bin:/home/ian/bin:/sbin:/usr/local/Trolltech/Qt-4.4.3/bin
    PERLDOC_PAGER=less
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @ig3

<details><summary>0002-Change-sense-of-test-for-HAS_GNULIBC.patch</summary>

```diff
From e8a24d4237ebc5ffc97de3051b883f8f6b9e8ef3 Mon Sep 17 00:00:00 2001
From: Ian Goodacre <Ian.Goodacre@xtra.co.nz>
Date: Mon, 16 Mar 2009 16:12:27 +1300
Subject: [PATCH 2/2] Change sense of test for HAS_GNULIBC

Time::Piece was reporting timezone offset incorrectly as follows:
$ ~/localperl/bin/perl -MTime::Piece -e '$t = Time::Piece->localtime; print $t->strftime("%z") . "\n"'
+0000
---
 ext/Time-Piece/Piece.xs |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/ext/Time-Piece/Piece.xs b/ext/Time-Piece/Piece.xs
index e20b6d4..e65f20c 100644
--- a/ext/Time-Piece/Piece.xs
+++ b/ext/Time-Piece/Piece.xs
@@ -23,7 +23,7 @@ extern "C" {
  * support is added and NETaa14816 is considered in full.
  * It does not address tzname aspects of NETaa14816.
  */
-#if !defined(HAS_GNULIBC)
+#if defined(HAS_GNULIBC)
 # ifndef STRUCT_TM_HASZONE
 #    define STRUCT_TM_HASZONE
 # else
-- 
1.6.1


```
</details>

**Comment by p5pRT** [comment]
### From @ig3

Attached is a patch to add a test for correct results from tzoffset and 
strftime\("%z"\)\.

Regards\,
Ian Goodacre

**Comment by p5pRT** [comment]
### From @ig3

<details><summary>0002-Added-test-for-tzoffset-and-strftime-z.patch</summary>

```diff
From a4da064be697cd6c59bb9f18c280bca4e6416eec Mon Sep 17 00:00:00 2001
From: Ian Goodacre <Ian.Goodacre@xtra.co.nz>
Date: Mon, 16 Mar 2009 18:10:37 +1300
Subject: [PATCH 2/2] Added test for tzoffset and strftime("%z")

---
 ext/Time-Piece/t/02core.t |    9 ++++++++-
 1 files changed, 8 insertions(+), 1 deletions(-)

diff --git a/ext/Time-Piece/t/02core.t b/ext/Time-Piece/t/02core.t
index d020558..fb43038 100644
--- a/ext/Time-Piece/t/02core.t
+++ b/ext/Time-Piece/t/02core.t
@@ -1,4 +1,4 @@
-use Test::More tests => 93;
+use Test::More tests => 95;
 
 my $is_win32 = ($^O =~ /Win32/);
 my $is_qnx = ($^O eq 'qnx');
@@ -47,6 +47,13 @@ cmp_ok($t->datetime, 'eq','2000-02-29T12:34:56');
 cmp_ok($t->daylight_savings, '==', 0);
 
 # ->tzoffset?
+{
+    local $ENV{TZ} = "EST";
+    my $lt = localtime;
+    cmp_ok(scalar($lt->tzoffset), 'eq', '-18000');
+    cmp_ok($lt->strftime("%z"), 'eq', '-0500');
+}
+
 cmp_ok(($t->julian_day / 2451604.0243 ) - 1, '<', 0.001);
 cmp_ok(($t->mjd        /   51603.52426) - 1, '<', 0.001);
 cmp_ok($t->week, '==', 9);
-- 
1.6.1


```
</details>

**Comment by p5pRT** [comment]
### From @smpeters

On Sun Mar 15 22&#8203;:14&#8203;:04 2009\, iang wrote&#8203;:
> Attached is a patch to add a test for correct results from tzoffset and 
> strftime\("%z"\)\.
> 
> Regards\,
> Ian Goodacre


Thanks\!  Both your patches were applied with the following 
change\.\.\.http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/036055ae

Steve Peters
steve@&#8203;fisharerojo\.org

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @craigberry

On Wed\, Mar 25\, 2009 at 8&#8203;:43 AM\, Steve Peters via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Sun Mar 15 22&#8203;:14&#8203;:04 2009\, iang wrote&#8203;:
>> Attached is a patch to add a test for correct results from tzoffset and
>> strftime\("%z"\)\.
>>
>> Regards\,
>> Ian Goodacre
>
>
> Thanks\!  Both your patches were applied with the following
> change\.\.\.http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/036055ae

This new test was failing on multiple platforms\.  I've taken a swing
at fixing it here&#8203;:

http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/d82dd1b

You can see whether I made it better or worse by following test reports here&#8203;:

http&#8203;://www\.nntp\.perl\.org/group/perl\.daily\-build\.reports/

**Comment by p5pRT** [comment]
### From @craigberry

On Fri\, Mar 27\, 2009 at 9&#8203;:41 AM\, Craig A\. Berry \<craig\.a\.berry@&#8203;gmail\.com> wrote&#8203;:
> On Wed\, Mar 25\, 2009 at 8&#8203;:43 AM\, Steve Peters via RT
> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>> On Sun Mar 15 22&#8203;:14&#8203;:04 2009\, iang wrote&#8203;:
>>> Attached is a patch to add a test for correct results from tzoffset and
>>> strftime\("%z"\)\.
>>>
>>> Regards\,
>>> Ian Goodacre
>>
>>
>> Thanks\!  Both your patches were applied with the following
>> change\.\.\.http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/036055ae
>
> This new test was failing on multiple platforms\.  I've taken a swing
> at fixing it here&#8203;:
>
> http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/d82dd1b
>
> You can see whether I made it better or worse by following test reports here&#8203;:
>
> http&#8203;://www\.nntp\.perl\.org/group/perl\.daily\-build\.reports/

I seem not to have made things worse\, but all is not entirely well
yet\.  Both of the new tests still fail on Windows\.  The first one
still fails on HP\-UX and AIX \(but no longer the second one after
changing %z to %Z\, so that's progress\)\.

But aside from the test failures\, something isn't quite right in this
chunk of the original patch&#8203;:

\-\-\- a/ext/Time\-Piece/Piece\.xs
\+\+\+ b/ext/Time\-Piece/Piece\.xs
@&#8203;@&#8203; \-23\,7 \+23\,7 @&#8203;@&#8203; extern "C" \{
&nbsp;  \* support is added and NETaa14816 is considered in full\.
&nbsp;  \* It does not address tzname aspects of NETaa14816\.
&nbsp;  \*/
\-\#if \!defined\(HAS\_GNULIBC\)
\+\#if defined\(HAS\_GNULIBC\)
 \# ifndef STRUCT\_TM\_HASZONE
 \#    define STRUCT\_TM\_HASZONE
 \# else
\[end of patch\]

What that does is select whether or not we call an init\_tm\(\) function
to initialize members of the tm struct that don't exist on some
platforms\.  Time&#8203;::Piece 1\.09 had "defined\(HAS\_GNULIBC\)"\, but 1\.10
changed it to "\!defined\(HAS\_GNULIBC\)" and now we've just switched it
back again in blead\.  That's not a sign that we know what we're doing
here\.

At one time\, whether you were using GNU libc or not may have been a
good enough poor man's approximation of whether your tm struct had
tm\_zone\.  But it was never a sustainable assumption that whatever GNU
libc is doing is the exact opposite of what the rest of the world is
doing\, and the only reason Time&#8203;::Piece still even has this code is
that it's a very old and unmaintained fork of some code that used to
be in POSIX\.xs but moved to util\.c eight years ago&#8203;:

http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/e72cf795

Ironically\, Jarkko's commit message indicates it was moved
specifically to make way for Time&#8203;::Piece in the core\.  So in blead
I've now \(finally\) made us take advantage of that&#8203;:

http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/d4d72fe

Not sure if that will make any difference in the remaining test failures\.

**Comment by p5pRT** [comment]
### From @craigberry

On Sun\, Mar 29\, 2009 at 8&#8203;:08 PM\, Ian Goodacre \<Ian\.Goodacre@&#8203;xtra\.co\.nz> wrote&#8203;:

> I note that the test is now a test of format %Z \(timezone name or
> abbreviation\) rather than %z \(timezone as hour offset from GMT\)\.

> At least on my system \(Linux 2\.6\.18\-92\.el5 \#1 SMP Tue Jun 10 18&#8203;:49&#8203;:47 EDT
> 2008 i686 i686 i386 GNU/Linux\) struct tm has two fields&#8203;: tm\_gmtoff \(or
> \_\_tm\_gmtoff\) for seconds east of UTC and tm\_zone \(or \_\_tm\_zone\) for timezone
> abbreviation\. The strftime\(\) function returns the value of the former for %z
> and the latter for %Z\. They appear to be quite independent of each other\,
> except to the extent that software that sets one sets the other
> consistently\.
>
> Perhaps the failure of %z to return the expected result is a real fault that
> does not correlate with whether %Z returns a correct result\, in which case
> it would be good to include both tests rather than one or the other\, and
> carry on to find out why %z is failing on so many systems\.

The problem with %z is that it's not in the standard and several
platforms don't support it&#8203;:

http&#8203;://www\.opengroup\.org/onlinepubs/7990989775/xsh/strftime\.html

> Also I note that perldoc POSIX says&#8203;: "The "Z" specifier is notoriously
> unportable since the names of timezones are non\-standard\. Sticking to the
> numeric specifiers is the safest route\." In which case testing %Z may prove
> problematic\, depending on locale\. I assume that in this case %z would be the
> relevant numeric specifier\. Maybe EST at least is broadly independent of
> locale\.

Hopefully EST will work everywhere\, but note that it is not valid to
specify a timezone name in the TZ environment variable without also
specifying an offset\, as I mentioned here&#8203;:

http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2009/03/msg145358\.html

> I had a look at the test reports but I only see "38" reported\. I am assuming
> 38 is the number of the failed test \- I'm not very familiar with this\.

Yes\, 38 is the number of your first test \(tzoffset\) and 39 of the
second \(strftime\("%Z"\)\)\.

> I am
> curious to know what was returned for %z on the failing systems\.  Is there
> some way to get more detail out of the smoke reports?

Not directly from smoke reports\, but there have been some results
posted in this and/or a parallel thread\.  Some platforms return
nothing for %z and I think I saw one that considers it an alias of %Z\.
 Basically it doesn't seem that it exists or means the same thing in
enough places to be useful for cross\-platform testing\.

> Please note that I only know enough to be dangerous so you should feel free
> to ignore my blathering\.

Everything I know I've learned by Googling\, reading the relevant
standards\, watching the smoke reports\, and reading replies on the
mailing list over the last couple of days\.
